R/data.table.R

```diff
@@ -1942,9 +1942,42 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
       X[[j]] = xj
     }
   }
-  X = unlist(X, recursive = FALSE, use.names = FALSE)
-  dim(X) <- c(n, length(X)/n)
-  dimnames(X) <- list(rownames.value, unlist(collabs, use.names = FALSE))
+  
+  # If any columns of x are multi-column, which can only occur if x is
+  # constructed incorrectly (see test 2074.07), use the old non-C
+  # method of converting to a matrix.
+  col.classes = sapply(X, class)
+  if (any(col.classes %in% c("matrix", "data.frame", "data.table"))) { 
+    X = unlist(X, recursive = FALSE, use.names = FALSE)
+    dim(X) = c(n, length(X)/n)
+    dimnames(X) = list(rownames.value, unlist(collabs, use.names = FALSE))
+    return(X)
+  } 
+  
+  # convert columns to common class before handing over to C
+  class.order = c("logical"=1, "integer"=2, "numeric"=3, "complex"=4,
+                  "character"=5, "raw"=6, "list"=7)
+  if (length(unique(col.classes)) > 1) { 
+    target.class = names(which.max(class.order[unique(col.classes)]))
+    if (target.class == "raw") target.class = "character" # to match behaviour of as.matrix.data.frame
+    for (col in which(col.classes != target.class)) {
```

Comment:
Instead of going column-by-column, would there be any performance benefit to using `lapply` here instead, like

```
idx = col.classes != target.class
X[idx] = lapply(X[idx], as, target.class)
```

?

Reply:
Alternatively, once we've got a `list` and don't have to worry about the nested dimensions bit, we know the size of the output matrix right? Can we use `[<-` in a loop and make R handle the coercion? Being sure to initialize the matrix with the correct output type...

Reply:
I've given that a try to simplify things but `as` has weird behaviour. Given an integer column, if I try `X[[idx]] = as(X[[idx]], "numeric")` the column remains integer type in R, but is recognised as a REALSXP in C, so the resulting matrix ends up with garbage values.